### PR TITLE
enhance mosn smooth upgrade for iusse #1397

### PR DESCRIPTION
**Pull request by alpha-baby**
### Issues associated with this PR

Your PR should present related issues you want to solve.

### Solutions
You should show your solutions about the issues in your PR, including the overall solutions, details and the changes. At this time, Chinese is allowed to describe these.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

**Comment by nejisama** [comment]
please add a test report, include :

**Comment by alpha-baby** [comment]
> 1. smooth upgrade without InheritOldMosnconfig, which will be uograded success

new mosn and old mosn use static config file：
```json
{
	"close_graceful" : false,
	"inherit_old_mosnconfig": false,
	"servers":[
		{
			"default_log_path":"stdout",
			"routers":[
				{
					"router_config_name":"server_router",
					"virtual_hosts":[{
						"name":"serverHost",
						"domains": ["*"],
						"routers": [
							{
								"match":{"prefix":"/"},
								"route":{"cluster_name":"serverCluster"}
							}
						]
					}]
				}
			],
			"listeners":[
				{
					"name":"serverListener",
					"address": "127.0.0.1:2046",
					"bind_port": true,
					"filter_chains": [{
						"filters": [
							{
								"type": "proxy",
								"config": {
									"downstream_protocol": "Http1",
									"upstream_protocol": "Http1",
									"router_config_name":"server_router"
								}
							}
						]
					}]
				}
			]
		}
	],
	"cluster_manager":{
		"clusters":[
			{
				"name":"serverCluster",
				"type": "SIMPLE",
				"lb_type": "LB_RANDOM",
				"max_request_per_conn": 1024,
				"conn_buffer_limit_bytes":32768,
				"hosts":[
					{"address":"127.0.0.1:8080"}
				]
			}
		]
	},
	"admin": {
		"address": {
			"socket_address": {
				"address": "0.0.0.0",
				"port_value": 34902
			}
		}
	}
}

```

**test pass**
**smooth upgrade success**


> 2. smooth upgrade with InheritOldMosnconfig, which will be upgraded success

new mosn and old mosn use static config file：
```json
{
	"close_graceful" : false,
	"inherit_old_mosnconfig": true,
	"servers":[
		{
			"default_log_path":"stdout",
			"routers":[
				{
					"router_config_name":"server_router",
					"virtual_hosts":[{
						"name":"serverHost",
						"domains": ["*"],
						"routers": [
							{
								"match":{"prefix":"/"},
								"route":{"cluster_name":"serverCluster"}
							}
						]
					}]
				}
			],
			"listeners":[
				{
					"name":"serverListener",
					"address": "127.0.0.1:2046",
					"bind_port": true,
					"filter_chains": [{
						"filters": [
							{
								"type": "proxy",
								"config": {
									"downstream_protocol": "Http1",
									"upstream_protocol": "Http1",
									"router_config_name":"server_router"
								}
							}
						]
					}]
				}
			]
		}
	],
	"cluster_manager":{
		"clusters":[
			{
				"name":"serverCluster",
				"type": "SIMPLE",
				"lb_type": "LB_RANDOM",
				"max_request_per_conn": 1024,
				"conn_buffer_limit_bytes":32768,
				"hosts":[
					{"address":"127.0.0.1:8080"}
				]
			}
		]
	},
	"admin": {
		"address": {
			"socket_address": {
				"address": "0.0.0.0",
				"port_value": 34902
			}
		}
	}
}

```

**test pass**
**smooth upgrade success**

> 3. old mosn config with InheritOldMosnconfig, new mosn config without InheritOldMosnconfig, which will be upgraded fail.

old mosn use config file, with InheritOldMosnconfig

```json
{
	"close_graceful" : false,
	"inherit_old_mosnconfig": true,
	"servers":[
		{
			"default_log_path":"stdout",
			"routers":[
				{
					"router_config_name":"server_router",
					"virtual_hosts":[{
						"name":"serverHost",
						"domains": ["*"],
						"routers": [
							{
								"match":{"prefix":"/"},
								"route":{"cluster_name":"serverCluster"}
							}
						]
					}]
				}
			],
			"listeners":[
				{
					"name":"serverListener",
					"address": "127.0.0.1:2046",
					"bind_port": true,
					"filter_chains": [{
						"filters": [
							{
								"type": "proxy",
								"config": {
									"downstream_protocol": "Http1",
									"upstream_protocol": "Http1",
									"router_config_name":"server_router"
								}
							}
						]
					}]
				}
			]
		}
	],
	"cluster_manager":{
		"clusters":[
			{
				"name":"serverCluster",
				"type": "SIMPLE",
				"lb_type": "LB_RANDOM",
				"max_request_per_conn": 1024,
				"conn_buffer_limit_bytes":32768,
				"hosts":[
					{"address":"127.0.0.1:8080"}
				]
			}
		]
	},
	"admin": {
		"address": {
			"socket_address": {
				"address": "0.0.0.0",
				"port_value": 34902
			}
		}
	}
}
```

new mosn use config file, without InheritOldMosnconfig

```json
{
	"close_graceful" : false,
	"inherit_old_mosnconfig": false,
	"servers":[
		{
			"default_log_path":"stdout",
			"routers":[
				{
					"router_config_name":"server_router",
					"virtual_hosts":[{
						"name":"serverHost",
						"domains": ["*"],
						"routers": [
							{
								"match":{"prefix":"/"},
								"route":{"cluster_name":"serverCluster"}
							}
						]
					}]
				}
			],
			"listeners":[
				{
					"name":"serverListener",
					"address": "127.0.0.1:2046",
					"bind_port": true,
					"filter_chains": [{
						"filters": [
							{
								"type": "proxy",
								"config": {
									"downstream_protocol": "Http1",
									"upstream_protocol": "Http1",
									"router_config_name":"server_router"
								}
							}
						]
					}]
				}
			]
		}
	],
	"cluster_manager":{
		"clusters":[
			{
				"name":"serverCluster",
				"type": "SIMPLE",
				"lb_type": "LB_RANDOM",
				"max_request_per_conn": 1024,
				"conn_buffer_limit_bytes":32768,
				"hosts":[
					{"address":"127.0.0.1:8080"}
				]
			}
		]
	},
	"admin": {
		"address": {
			"socket_address": {
				"address": "0.0.0.0",
				"port_value": 34902
			}
		}
	}
}
```

**test not pass**
**smooth not fail, Both the new mosn and the old mosn will always run, new mosn not will be exit**

new commit fix this bug.

> 4. old mosn config without InheritOldMosnconfig, new mosn config with InheritOldMosnconfig, which will be upgraded fail.

old mosn use config file, without inheritOldMosnconfig

```json
{
	"close_graceful" : false,
	"inherit_old_mosnconfig": false,
	"servers":[
		{
			"default_log_path":"stdout",
			"routers":[
				{
					"router_config_name":"server_router",
					"virtual_hosts":[{
						"name":"serverHost",
						"domains": ["*"],
						"routers": [
							{
								"match":{"prefix":"/"},
								"route":{"cluster_name":"serverCluster"}
							}
						]
					}]
				}
			],
			"listeners":[
				{
					"name":"serverListener",
					"address": "127.0.0.1:2046",
					"bind_port": true,
					"filter_chains": [{
						"filters": [
							{
								"type": "proxy",
								"config": {
									"downstream_protocol": "Http1",
									"upstream_protocol": "Http1",
									"router_config_name":"server_router"
								}
							}
						]
					}]
				}
			]
		}
	],
	"cluster_manager":{
		"clusters":[
			{
				"name":"serverCluster",
				"type": "SIMPLE",
				"lb_type": "LB_RANDOM",
				"max_request_per_conn": 1024,
				"conn_buffer_limit_bytes":32768,
				"hosts":[
					{"address":"127.0.0.1:8080"}
				]
			}
		]
	},
	"admin": {
		"address": {
			"socket_address": {
				"address": "0.0.0.0",
				"port_value": 34902
			}
		}
	}
}

```

new mosn use config file, with inheritOldMosnconfig

```json
{
	"close_graceful" : false,
	"inherit_old_mosnconfig": true,
	"servers":[
		{
			"default_log_path":"stdout",
			"routers":[
				{
					"router_config_name":"server_router",
					"virtual_hosts":[{
						"name":"serverHost",
						"domains": ["*"],
						"routers": [
							{
								"match":{"prefix":"/"},
								"route":{"cluster_name":"serverCluster"}
							}
						]
					}]
				}
			],
			"listeners":[
				{
					"name":"serverListener",
					"address": "127.0.0.1:2046",
					"bind_port": true,
					"filter_chains": [{
						"filters": [
							{
								"type": "proxy",
								"config": {
									"downstream_protocol": "Http1",
									"upstream_protocol": "Http1",
									"router_config_name":"server_router"
								}
							}
						]
					}]
				}
			]
		}
	],
	"cluster_manager":{
		"clusters":[
			{
				"name":"serverCluster",
				"type": "SIMPLE",
				"lb_type": "LB_RANDOM",
				"max_request_per_conn": 1024,
				"conn_buffer_limit_bytes":32768,
				"hosts":[
					{"address":"127.0.0.1:8080"}
				]
			}
		]
	},
	"admin": {
		"address": {
			"socket_address": {
				"address": "0.0.0.0",
				"port_value": 34902
			}
		}
	}
}

```

**test pass**
**smooth upgrade fail**

**Comment by alpha-baby** [comment]
@nejisama

**Comment by alpha-baby** [comment]
@nejisama 新的提交 使用 time sleep 这样解决比较垃圾，但是我目前也没有想到什么更好的方式了，你们也想想有什么更好的方式不

File: pkg/mosn/starter.go [hunk 1]
```diff
@@ -202,7 +217,8 @@ func (m *Mosn) beforeStart() {
 		if err := store.StartService(m.inheritListeners); err != nil {
 			log.StartLogger.Fatalf("[mosn] [NewMosn] start service failed: %v,  exit", err)
 		}
-
+		// wait old mosn send mosnconfig
+		time.Sleep(time.Second * 11)
```
Comment (taoyuanyuan) [inline pkg/mosn/starter.go:221]: 这儿为啥要sleep哟？
GetInheritConfig的时候不是已经拿到了吗

Comment (alpha-baby) [reply pkg/mosn/starter.go:221]: 在测试用例的情况：

>3. old mosn config with InheritOldMosnconfig, new mosn config without InheritOldMosnconfig, which will be upgraded fail.

old mosn 会调用SendMosnconfig 但是 new mosn 并不会接收，然而old mosn 或有重试，这里的sleep 就是为了等待old mosn 重试完成，然后这样 new mosn 才会感知到old mosn 在发送数据的过程中失败了

Comment (nejisama) [reply pkg/mosn/starter.go:221]: 这里Slepp 11 秒肯定是不行的，这样写了以后。启动至少要11s才能往后走了

Comment (taoyuanyuan) [reply pkg/mosn/starter.go:221]: 再加一次ack，old mosn的 listenSockConn 收到 new mosn的ack之后，在回复一个就行了吧

Comment (taoyuanyuan) [reply pkg/mosn/starter.go:221]:
```
		// notify old mosn to transfer connection
		if _, err := m.listenSockConn.Write([]byte{0}); err != nil {
			log.StartLogger.Fatalf("[mosn] [NewMosn] graceful failed, exit")
		}
```
在这之后再来一次 Read old mosn的 ack

Comment (alpha-baby) [reply pkg/mosn/starter.go:221]: 这样感觉似乎可以，我后面写代码试试

**Comment by alpha-baby** [comment]
> please add a test report(do it manually it's ok), include :
> 
> 1. smooth upgrade without InheritOldMosnconfig, which will be uograded success
> 2. smooth upgrade with InheritOldMosnconfig, which will be upgraded success
> 3. old mosn config with InheritOldMosnconfig, new mosn config without InheritOldMosnconfig, which will be upgraded fail.
> 4. old mosn config without InheritOldMosnconfig, new mosn config with InheritOldMosnconfig, which will be upgraded fail.

all test case pass @nejisama

File: pkg/server/reconfigure.go [hunk 2]
```diff
@@ -89,17 +89,24 @@ func reconfigure(start bool) {
 	if listenSockConn, err = sendInheritListeners(); err != nil {
 		return
 	}
+	// 0 means success, not 0 means failure
+	var reponseFlag uint8 = 0
```
Comment (taoyuanyuan) [inline pkg/server/reconfigure.go:93]: 不需要搞一个resonoseflag，跟之前一样的方式就行了，直接close或者发一个包，就验证error或者n== 1就行了

File: pkg/mosn/starter.go [hunk 3]
```diff
@@ -202,11 +218,22 @@ func (m *Mosn) beforeStart() {
 		if err := store.StartService(m.inheritListeners); err != nil {
 			log.StartLogger.Fatalf("[mosn] [NewMosn] start service failed: %v,  exit", err)
 		}
-
 		// notify old mosn to transfer connection
 		if _, err := m.listenSockConn.Write([]byte{0}); err != nil {
 			log.StartLogger.Fatalf("[mosn] [NewMosn] graceful failed, exit")
 		}
+		// wait old mosn ack
+		m.listenSockConn.SetReadDeadline(time.Now().Add(1 * time.Minute))
```
Comment (taoyuanyuan) [inline pkg/mosn/starter.go:226]: 这个就是兜底，正常应该是马上就能读取的，设置个3s估计就行了

File: pkg/mosn/starter.go [hunk 4]
```diff
@@ -65,6 +66,9 @@ func NewMosn(c *v2.MOSNConfig) *Mosn {
 	initializePidFile(c.Pid)
 	initializeTracing(c.Tracing)
 	initializePlugin(c.Plugin.LogBase)
+	if c.InheritOldMosnconfig {
+		server.EnableInheritOldMosnconfig()
```
Comment (taoyuanyuan) [inline pkg/mosn/starter.go:70]: server.EnableInheritOldMosnconfig(c.InheritOldMosnconfig)

File: pkg/server/reconfigure.go [hunk 5]
```diff
@@ -82,12 +89,31 @@ func reconfigure(start bool) {
 	if listenSockConn, err = sendInheritListeners(); err != nil {
 		return
 	}
+	// 0 means success, not 0 means failure
```
Comment (taoyuanyuan) [inline pkg/server/reconfigure.go:92]: unused comment

**Review by taoyuanyuan — APPROVED** [review]
LGTM

**Comment by alpha-baby** [comment]
> please add a test report(do it manually it's ok), include :
> 
> 1. smooth upgrade without InheritOldMosnconfig, which will be uograded success
> 2. smooth upgrade with InheritOldMosnconfig, which will be upgraded success
> 3. old mosn config with InheritOldMosnconfig, new mosn config without InheritOldMosnconfig, which will be upgraded fail.
> 4. old mosn config without InheritOldMosnconfig, new mosn config with InheritOldMosnconfig, which will be upgraded fail.

**use test config file:** `examples/codes/http-sample/server.go`

## case 1:

### new mosn log

```log
2020-12-07 17:27:09,780 [INFO] [router] [Extend] [RegisterRouterRule] order is 1
2020-12-07 17:27:09,780 [INFO] [router] [Extend] [RegisterHandlerChain] order is 1
2020-12-07 17:27:09,781 [INFO] [config] processor added to configParsedCBMaps
2020-12-07 17:27:09,784 [INFO] [network] [ register pool factory] register protocol: Http1 factory
2020-12-07 17:27:09,784 [INFO] [network] [ register pool factory] register protocol: Http2 factory
2020-12-07 17:27:09,784 [INFO] [network] [ register pool factory] register protocol: X factory
2020-12-07 17:27:09,788 [INFO] load config from :  server_config.json
2020-12-07 17:27:09,789 [INFO] [mosn] [init tracing] disable tracing
2020-12-07 17:27:09,792 [INFO] [server] Get InheritListeners start
2020-12-07 17:27:09,793 [INFO] [server] Get InheritListeners Accept
2020-12-07 17:27:09,793 [INFO] [mosn] [NewMosn] active reconfiguring
2020-12-07 17:27:09,793 [INFO] [admin store] [mosn state] state changed to 2
2020-12-07 17:27:09,793 [INFO] load config from :  server_config.json
2020-12-07 17:27:09,794 [DEBUG] [server] [init] add server conn handler, server name = 
2020-12-07 17:27:09,794 [INFO] parsing listen config:tcp
2020-12-07 17:27:09,794 [INFO] [cluster] [cluster manager] [AddOrUpdatePrimaryCluster] cluster serverCluster updated
2020-12-07 17:27:09,794 [INFO] [upstream] [host set] update host, final host total: 1
2020-12-07 17:27:09,794 [INFO] [cluster] [primaryCluster] [UpdateHosts] cluster serverCluster update hosts: 1
2020-12-07 17:27:09,794 [INFO] [config] [parse listener] [tcp] inherit listener addr: 127.0.0.1:2046
2020-12-07 17:27:09,794 [INFO] [config] [parse listener] [tcp] inherit listener addr: 127.0.0.1:2047
2020-12-07 17:27:09,794 [INFO] mosn start xds client
2020-12-07 17:27:09,794 [WARN] [feature gate] feature XdsMtlsEnable is not enabled
2020-12-07 17:27:09,794 [WARN] [feature gate] feature PayLoadLimitEnable is not enabled
2020-12-07 17:27:09,794 [WARN] [feature gate] feature MultiTenant2020-12-07 17:27:09,794 [INFO] [server] [conn handler] [add listener] add listener: 127.0.0.1:2046
2020-12-07 17:27:09,794 [INFO] parsing listen config:tcp
2020-12-07 17:27:09,794 [INFO] [server] [conn handler] [add listener] add listener: 127.0.0.1:2047
2020-12-07 17:27:09,794 [INFO] [router] [virtualhost] [addRouteBase] add a new route rule
2020-12-07 17:27:09,794 [INFO] [router] [routers] [NewRouters] add route matcher default virtual host
2020-12-07 17:27:09,794 [INFO] [router] [routers_manager] [AddOrUpdateRouters] add router: server_router
2020-12-07 17:27:09,794 [INFO] xds client start
2020-12-07 17:27:09,794 [ERROR] StaticResources is null
2020-12-07 17:27:09,794 [WARN] fail to init xds config, skip xds: null point exception
2020-12-07 17:27:09,794 [INFO] [admin store] [add service] add server Mosn Admin Server
2020-12-07 17:27:09,794 [INFO] [network] [transfer] [server] TransferServer start
Mode is not enabled
2020-12-07 17:27:09,794 [WARN] [feature gate] feature auto_config is not enabled
2020-12-07 17:27:09,794 [INFO] mosn parse extend config
2020-12-07 17:27:09,794 [INFO] mosn prepare for start
2020-12-07 17:27:09,794 [INFO] [admin store] [start service] [inheritListener] inherit listener addr: [::]:34902
2020-12-07 17:27:09,794 [INFO] [admin store] [start service] start service Mosn Admin Server on [::]:34902
2020-12-07 17:27:09,794 [INFO] mosn start server
2020-12-07 17:27:10,795 [INFO] [server] [reconfigure] reconfigureHandler start
2020-12-07 17:28:49,800 [INFO] [network] [transfer] [server] TransferServer exit
2020-12-07 17:28:49,801 [INFO] [network] [transfer] [server] TransferServer listener conn.sock closed
2020-12-07 17:28:49,801 [INFO] [admin store] [mosn state] state changed to 1
```

### old mosn log

```log
// .....
2020-12-07 17:27:09,792 [INFO] [server] [reconfigure] reconfigureHandler Accept
2020-12-07 17:27:09,792 [INFO] [admin store] [mosn state] state changed to 3
2020-12-07 17:27:09,793 [DEBUG] [server] InheritListener fd: 12
2020-12-07 17:27:09,793 [DEBUG] [server] InheritListener fd: 13
2020-12-07 17:27:09,793 [DEBUG] [server] InheritListener fd: 14
2020-12-07 17:27:09,794 [INFO] [admin store] [stop service] clear all stored services
2020-12-07 17:27:09,794 [INFO] [admin store] [stop service] Mosn Admin Server
2020-12-07 17:27:09,794 [WARN] [admin store] [start service] start serve failed : Mosn Admin Server [::]:34902 http: Server closed
2020-12-07 17:27:12,796 [INFO] [server] StopConnection
2020-12-07 17:27:12,796 [INFO] [network] [listener start] [accept] listener serverListener stop accepting connections by deadline
2020-12-07 17:27:12,796 [INFO] [network] [listener start] [accept] listener serverListener2 stop accepting connections by deadline
2020-12-07 17:28:42,803 [INFO] [server] [reconfigure] process 55792 gracefully shutdown
```

## case 2

### new mosn log

```log
2020-12-07 17:32:43,796 [INFO] [router] [Extend] [RegisterRouterRule] order is 1
2020-12-07 17:32:43,797 [INFO] [router] [Extend] [RegisterHandlerChain] order is 1
2020-12-07 17:32:43,798 [INFO] [config] processor added to configParsedCBMaps
2020-12-07 17:32:43,801 [INFO] [network] [ register pool factory] register protocol: Http1 factory
2020-12-07 17:32:43,801 [INFO] [network] [ register pool factory] register protocol: Http2 factory
2020-12-07 17:32:43,801 [INFO] [network] [ register pool factory] register protocol: X factory
2020-12-07 17:32:43,804 [INFO] load config from :  server_config.json
2020-12-07 17:32:43,806 [INFO] [mosn] [init tracing] disable tracing
2020-12-07 17:32:43,807 [INFO] [server] Get InheritListeners start
2020-12-07 17:32:43,807 [INFO] [server] Get InheritListeners Accept
2020-12-07 17:32:43,808 [INFO] [mosn] [NewMosn] active reconfiguring
2020-12-07 17:32:43,808 [INFO] [admin store] [mosn state] state changed to 2
2020-12-07 17:32:43,808 [INFO] load config from :  server_config.json
2020-12-07 17:32:43,808 [INFO] [server] Get GetInheritConfig start
2020-12-07 17:32:44,812 [INFO] [server] Get GetInheritConfig Accept
2020-12-07 17:32:44,813 [INFO] [cluster] [cluster manager] [AddOrUpdatePrimaryCluster] cluster serverCluster updated
2020-12-07 17:32:44,813 [INFO] [upstream] [host set] update host, final host total: 1
2020-12-07 17:32:44,814 [INFO] [cluster] [primaryCluster] [UpdateHosts] cluster serverCluster update hosts: 1
2020-12-07 17:32:44,814 [INFO] [config] [parse listener] [tcp] inherit listener addr: 127.0.0.1:2046
2020-12-07 17:32:44,814 [INFO] [config] [parse listener] [tcp] inherit listener addr: 127.0.0.1:2047
2020-12-07 17:32:44,814 [INFO] mosn start xds client
2020-12-07 17:32:44,814 [WARN] [feature gate] feature XdsMtlsEnable is not enabled
2020-12-07 17:32:44,814 [WARN] [feature gate] feature PayLoadLimitEnable is not enabled
2020-12-07 17:32:44,814 [WARN] [feature gate] feature MultiTenantMode is not enabled
2020-12-07 17:32:44,814 [WARN] [feature gate] feature auto_config is not enabled
2020-12-07 17:32:44,814 [INFO] mosn parse extend config
2020-12-07 17:32:44,814 [INFO] mosn prepare for start
2020-12-07 17:32:44,814 [DEBUG] [server] [init] add server conn handler, server name = 
2020-12-07 17:32:44,814 [INFO] parsing listen config:tcp
2020-12-07 17:32:44,814 [INFO] [server] [conn handler] [add listener] add listener: 127.0.0.1:2046
2020-12-07 17:32:44,814 [INFO] parsing listen config:tcp
2020-12-07 17:32:44,814 [INFO] [server] [conn handler] [add listener] add listener: 127.0.0.1:2047
2020-12-07 17:32:44,814 [INFO] [router] [virtualhost] [addRouteBase] add a new route rule
2020-12-07 17:32:44,814 [INFO] [router] [routers] [NewRouters] add route matcher default virtual host
2020-12-07 17:32:44,814 [INFO] [router] [routers_manager] [AddOrUpdateRouters] add router: server_router
2020-12-07 17:32:44,814 [INFO] [admin store] [start service] [inheritListener] inherit listener addr: [::]:34902
2020-12-07 17:32:44,814 [INFO] [admin store] [start service] start service Mosn Admin Server on [::]:34902
2020-12-07 17:32:44,814 [INFO] mosn start server
2020-12-07 17:32:44,814 [INFO] xds client start
2020-12-07 17:32:44,814 [ERROR] StaticResources is null
2020-12-07 17:32:44,814 [WARN] fail to init xds config, skip xds: null point exception
2020-12-07 17:32:44,814 [INFO] [admin store] [add service] add server Mosn Admin Server
2020-12-07 17:32:44,815 [INFO] [network] [transfer] [server] TransferServer start
2020-12-07 17:32:45,819 [INFO] [server] [reconfigure] reconfigureHandler start
2020-12-07 17:34:24,817 [INFO] [network] [transfer] [server] TransferServer exit
2020-12-07 17:34:24,818 [INFO] [admin store] [mosn state] state changed to 1
2020-12-07 17:34:24,818 [INFO] [network] [transfer] [server] TransferServer listener conn.sock closed
```

### old mosn log

```log
2020-12-07 17:32:43,807 [INFO] [server] [reconfigure] reconfigureHandler Accept
2020-12-07 17:32:43,807 [INFO] [admin store] [mosn state] state changed to 3
2020-12-07 17:32:43,807 [DEBUG] [server] InheritListener fd: 12
2020-12-07 17:32:43,807 [DEBUG] [server] InheritListener fd: 13
2020-12-07 17:32:43,807 [DEBUG] [server] InheritListener fd: 14
2020-12-07 17:32:44,814 [INFO] [admin store] [stop service] clear all stored services
2020-12-07 17:32:44,814 [INFO] [admin store] [stop service] Mosn Admin Server
2020-12-07 17:32:44,814 [WARN] [admin store] [start service] start serve failed : Mosn Admin Server [::]:34902 http: Server closed
2020-12-07 17:32:47,818 [INFO] [server] StopConnection
2020-12-07 17:32:47,818 [INFO] [network] [listener start] [accept] listener serverListener2 stop accepting connections by deadline
2020-12-07 17:32:47,818 [INFO] [network] [listener start] [accept] listener serverListener stop accepting connections by deadline
2020-12-07 17:34:17,820 [INFO] [server] [reconfigure] process 57004 gracefully shutdown
```

## case 3

### new mosn log

```log
//.....
2020-12-07 17:38:50,166 [INFO] [server] Get InheritListeners start
2020-12-07 17:38:50,167 [INFO] [server] Get InheritListeners Accept
2020-12-07 17:38:50,167 [INFO] [mosn] [NewMosn] active reconfiguring
2020-12-07 17:38:50,167 [INFO] [admin store] [mosn state] state changed to 2
2020-12-07 17:38:50,167 [INFO] load config from :  server_config.json
2020-12-07 17:38:50,168 [DEBUG] [server] [init] add server conn handler, server name = 
2020-12-07 17:38:50,168 [INFO] parsing listen config:tcp
2020-12-07 17:38:50,168 [INFO] [server] [conn handler] [add listener] add listener: 127.0.0.1:2046
2020-12-07 17:38:50,168 [INFO] parsing listen config:tcp
2020-12-07 17:38:50,167 [INFO] [cluster] [cluster manager] [AddOrUpdatePrimaryCluster] cluster serverCluster updated
2020-12-07 17:38:50,167 [INFO] [upstream] [host set] update host, final host total: 1
2020-12-07 17:38:50,167 [INFO] [cluster] [primaryCluster] [UpdateHosts] cluster serverCluster update hosts: 1
2020-12-07 17:38:50,168 [INFO] [config] [parse listener] [tcp] inherit listener addr: 127.0.0.1:2046
2020-12-07 17:38:50,168 [INFO] [config] [parse listener] [tcp] inherit listener addr: 127.0.0.1:2047
2020-12-07 17:38:50,168 [INFO] mosn start xds client
2020-12-07 17:38:50,168 [WARN] [feature gate] feature PayLoadLimitEnable is not enabled
2020-12-07 17:38:50,168 [WARN] [feature gate] feature MultiTenantMode is not enabled
2020-12-07 17:38:50,168 [WARN] [feature gate] feature auto_config is not enabled
2020-12-07 17:38:50,168 [WARN] [feature gate] feature XdsMtlsEnable is not enabled
2020-12-07 17:38:50,168 [INFO] mosn parse extend config
2020-12-07 17:38:50,168 [INFO] mosn prepare for start
2020-12-07 17:38:50,168 [INFO] [admin store] [start service] [inheritListener] inherit listener addr: [::]:34902
2020-12-07 17:38:50,168 [INFO] [admin store] [start service] start service Mosn Admin Server on [::]:34902
2020-12-07 17:38:50,168 [INFO] [server] [conn handler] [add listener] add listener: 127.0.0.1:2047
2020-12-07 17:38:50,168 [INFO] [router] [virtualhost] [addRouteBase] add a new route rule
2020-12-07 17:38:50,168 [INFO] [router] [routers] [NewRouters] add route matcher default virtual host
2020-12-07 17:38:50,168 [INFO] [router] [routers_manager] [AddOrUpdateRouters] add router: server_router
2020-12-07 17:38:50,168 [INFO] xds client start
2020-12-07 17:38:50,168 [ERROR] StaticResources is null
2020-12-07 17:38:50,168 [WARN] fail to init xds config, skip xds: null point exception
2020-12-07 17:38:50,168 [INFO] [admin store] [add service] add server Mosn Admin Server
2020-12-07 17:38:53,173 [FATAL] [mosn] [NewMosn] ack graceful failed, exit, error: read unix listen.sock->: i/o timeout n: 0, buf[0]: 0

```

### old mosn log

```log
2020-12-07 17:38:50,166 [INFO] [server] [reconfigure] reconfigureHandler Accept
2020-12-07 17:38:50,166 [INFO] [admin store] [mosn state] state changed to 3
2020-12-07 17:38:50,167 [DEBUG] [server] InheritListener fd: 12
2020-12-07 17:38:50,167 [DEBUG] [server] InheritListener fd: 13
2020-12-07 17:38:50,167 [DEBUG] [server] InheritListener fd: 14
2020-12-07 17:39:00,206 [ERROR] [server] SendInheritConfig Dial unix failed dial unix mosnconfig.sock: connect: no such file or directory
2020-12-07 17:39:00,207 [ERROR] [mosn.common.reconfigure_failed] [old mosn] [SendInheritConfig] new mosn start failed
2020-12-07 17:39:00,207 [INFO] [admin store] [mosn state] state changed to 1
```

## case 4

### new mosn log

```log
//...
2020-12-07 17:42:24,118 [INFO] [server] Get InheritListeners start
2020-12-07 17:42:24,118 [INFO] [server] Get InheritListeners Accept
2020-12-07 17:42:24,118 [INFO] [mosn] [NewMosn] active reconfiguring
2020-12-07 17:42:24,118 [INFO] [admin store] [mosn state] state changed to 2
2020-12-07 17:42:24,119 [INFO] load config from :  server_config.json
2020-12-07 17:42:24,119 [INFO] [server] Get GetInheritConfig start
2020-12-07 17:42:34,123 [ERROR] [server] GetInheritConfig Accept error :accept unix mosnconfig.sock: i/o timeout
2020-12-07 17:42:34,123 [FATAL] [mosn] [NewMosn] GetInheritConfig failed, exit

```

### old mosn log

```log
//....
2020-12-07 17:42:24,118 [INFO] [server] [reconfigure] reconfigureHandler Accept
2020-12-07 17:42:24,118 [INFO] [admin store] [mosn state] state changed to 3
2020-12-07 17:42:24,118 [DEBUG] [server] InheritListener fd: 12
2020-12-07 17:42:24,118 [DEBUG] [server] InheritListener fd: 13
2020-12-07 17:42:24,118 [DEBUG] [server] InheritListener fd: 14
2020-12-07 17:42:34,123 [ERROR] [mosn.common.reconfigure_failed] [old mosn] [read ack] new mosn start failed
2020-12-07 17:42:34,124 [INFO] [admin store] [mosn state] state changed to 1
```

# all test case pass

**Review by nejisama — APPROVED** [review]

**Pull request merged**